source/tests/SGM/SGMMonochromatorInfoTestView.cpp

```diff
@@ -0,0 +1,428 @@
+
+            warningString.append(QString("\n%1").arg(message));
+        }
+
+        warningsTextEdit_->setText(warningString);
+    }
+
+}
+
+void SGMMonochromatorInfoTestView::onMonoErrorCountChanged(int errorCount)
+{
+    if(errorCount == 0) {
+        errorsTextEdit_->setVisible(false);
+    } else {
+
+        errorsTextEdit_->setVisible(true);
+        QString errorString;
+
+        foreach(QString message, sgmMonochromatorInfo_->errorMessages()) {
+
+            errorString.append(QString("\n%1").arg(message));
+        }
+
+        errorsTextEdit_->setText(errorString);
+    }
+}
+
+void SGMMonochromatorInfoTestView::onGratingAngleSpinBoxChanged()
+{
+    double value = gratingAngleSpinBox_->value();
+    if(qAbs(value - sgmMonochromatorInfo_->gratingAngle()) > 0.001 ||
+            sgmMonochromatorInfo_->hasErrors()) {
+        sgmMonochromatorInfo_->setGratingAngle(value);
+    }
+}
+
+void SGMMonochromatorInfoTestView::onGratingSelectionModeComboBoxChanged(int index)
+{
+    gratingTranslationComboBox_->setEnabled(index == 3);
+    if(index != 3) {
+
+        SGMMonochromatorInfo::GratingTranslationOptimizationMode correspondingOptimizeMode;
+        switch(index) {
+        case 0:
+            correspondingOptimizeMode = SGMMonochromatorInfo::OptimizeFlux;
+            break;
+        case 1:
+            correspondingOptimizeMode = SGMMonochromatorInfo::OptimizeResolution;
+            break;
+        case 2:
+        default:
+            correspondingOptimizeMode = SGMMonochromatorInfo::OptimizeMinimalMovement;
+            break;
+        }
+
+
+        sgmMonochromatorInfo_->requestEnergy(this->energySpinBox_->value(), correspondingOptimizeMode);
+    }
+}
+
+void SGMMonochromatorInfoTestView::onGratingTranslationComboBoxChanged(int index)
+{
+    SGMGratingSupport::GratingTranslation correspondingGrating;
+    switch(index) {
+    case 0:
+        correspondingGrating = SGMGratingSupport::LowGrating;
+        break;
+    case 1:
+        correspondingGrating = SGMGratingSupport::MediumGrating;
+        break;
+    case 2:
+        correspondingGrating = SGMGratingSupport::HighGrating;
+        break;
+    default:
+        correspondingGrating = SGMGratingSupport::UnknownGrating;
+    }
+
+    if(correspondingGrating != SGMGratingSupport::UnknownGrating &&
+            correspondingGrating != sgmMonochromatorInfo_->gratingTranslation()) {
+
+        sgmMonochromatorInfo_->setGratingTranslation(correspondingGrating);
+    }
+}
+
+void SGMMonochromatorInfoTestView::onUndulatorTrackingCheckBoxChanged(bool value)
+{
+    if(sgmMonochromatorInfo_->isUndulatorTracking() != value) {
+        sgmMonochromatorInfo_->setUndulatorTracking(value);
+    }
+}
+
+void SGMMonochromatorInfoTestView::onUndulatorPositionSpinBoxChanged()
+{
+    double value = undulatorPositionSpinBox_->value();
+    if(qAbs(sgmMonochromatorInfo_->undulatorPosition() - value) > 0.001 ||
+            sgmMonochromatorInfo_->hasErrors()) {
```

Comment:
Same deal with these guys.  If this is on purpose and intended then you can disregard.